lib/sendgrid/mail.rb

```diff
@@ -28,34 +33,46 @@ def add_attachment(path, name = nil)
     end
 
     def to_h
-      payload = {
-        :from        => @from,
-        :fromname    => @from_name,
-        :subject     => @subject,
-        :to          => @to,
-        :toname      => @to_name,
-        :date        => @date,
-        :replyto     => @reply_to,
-        :cc          => @cc,
-        :bcc         => @bcc,
-        :text        => @text,
-        :html        => @html,
+      payload = {}
+
+      PAYLOAD_PARAMS.each do |payload_param|
+        payload[payload_mapping(payload_param)] = self.send(payload_param)
+      end
+
+      payload.merge!({
         :'x-smtpapi' => smtpapi_json,
         :files       => ({} unless @attachments.empty?)
-      }.reject {|k,v| v.nil?}
+      })
+
+      payload.reject! {|k,v| v.nil?}
+
+      assign_missing_to(payload)
+      extract_attachments(payload)
```

Comment:
`:files` will be `nil` if `@attachments` is empty. The `merge!` call above assigns `:files` to `({} unless @attachments.empty?)`. Then this call does nothing if `@attachments` is empty. So `:files` can be `nil`. I don't know if that will cause errors down the line.

Anyway, I think we're in agreement on writing `extract_attachments` to functionally return the attachments, instead of modify the payload in place.